Added the `translate` module for MT

alvations opened the pull request:
- Removed the un-used import in `align/gdfa.py`
- Added the translate module with `model.py` to load pre-calculated phrase table and language models and
- added `stack_decoder.py` to perform stack_decoding for Statistical Machine Translation (SMT)

stevenbird commented on nltk/translate/models.py:
```diff
@@ -0,0 +1,201 @@
+class LanguageModel:
+    """
+    This class implements an Ngram language model object that reads an ARPA 
+    language model file usinand allow users to retrieve the probability of a 
```
typo?

stevenbird commented on nltk/translate/models.py:
```diff
@@ -0,0 +1,201 @@
+        lang_model[('<unk>',)] = (float(-100), 0)
+    return lang_model
+
+class LanguageModel:
```
Ultimately, this class will probably belong in a different package.

stevenbird commented on nltk/translate/stack_decoder.py:
```diff
@@ -0,0 +1,141 @@
+    :type sent: list
+    :param sent: list of tokens from the source language sentence.
+
+    :type tm: 
```
Missing type

stevenbird commented on nltk/translate/stack_decoder.py:
```diff
@@ -0,0 +1,141 @@
+    :type tm: 
+    :param tm: Translation model from the phrase table.
+    
+    :type lm:
```
Missing type

stevenbird commented:
@alvations please give me the URL for the models.

alvations commented:
@stevenbird sorry for the late reply, I was busy with something else yesterday.

I've fixed the typos and added the types for the decoder input parameters. 

The language model sample file is in the `/sample-models/lm/europarl.srilm.gz` and the phrase table sample file is in `/sample-models/phrase-model/phrase-table` from
http://www.statmt.org/moses/download/sample-models.tgz . Additionally, you would need to compress the `/sample-models/phrase-model/phrase-table` file with gzip into `.gz` format

About the sample files, the Moses toolkit developers just updated their documentation and uploaded a different version, where the language model is gzipped and the phrase table isn't. I think it's still useful to `read_phrase_table()` with `gzip.open()` since real data does that. 

Regarding the models.py, I too agree that it should go somewhere else.

stevenbird commented:
We can make a compressed file `sample-models.zip` available in the nltk data collection. It won't matter that the phrase table isn't compressed. If I do that, would you be happy to modify your code to work with that version? I think it will just make it easier for us to deal with future versions of the Moses data.

alvations commented:
No problem, i can change the code to suit the non-compressed file.

alvations commented:
@stevenbird , I've updated the code and here's a link to the `sample-model.zip` with the `phrase-table` file and `europarl.srilm.gz`: https://db.tt/OlzKtfgE . Possibly the path to the models needs to be changed in the doctest code snippets.

stevenbird commented:
The data is now in nltk_data, cf https://github.com/nltk/nltk_data/commit/6d23de5c4c3383e90f0d7eb44a98faf3fd3de50c

stevenbird commented:
Would you please update your pull request to access the language models using a method in `nltk.data`, and make this conform to the [requirements for NLTK package structure](https://github.com/nltk/nltk/wiki/Package-Structure).

alvations commented:
I'll work on the update with the package requirements and push it up later this week.

stevenbird commented:
OK, thanks.

alvations commented:
@stevenbird , sorry for the delayed response. Now the code is using the zip file from `nltk_data`. I did some reshuffling of the functions to suit the NLTK package requirements.

stevenbird commented on nltk/translate/util.py:
```diff
@@ -0,0 +1,170 @@
+# -*- coding: utf-8 -*-
+# Natural Language Toolkit: Machine Translation Utilities
+#
+# Copyright (C) 2001-2013 NLTK Project
```
@alvations would you please fix the copyright year range, and add author info?

alvations replied:
fixed.

stevenbird commented on nltk/translate/util.py:
```diff
@@ -0,0 +1,170 @@
+    >>> "/".join(model_filename.split('/')[-5:])
+    'nltk_data/models/moses_sample/lm/europarl.srilm.gz'
+    
+    :type approach: str
```
@alvations would you please move the parameter declarations to the bottom of the docstring?

alvations replied:
fixed.

stevenbird commented on nltk/translate/api.py:
```diff
@@ -0,0 +1,101 @@
+
+from util import read_phrase_table, read_lang_model
+
+class LanguageModel:
```
@alvations, please explicitly subclass from object, since we're still supporting Python 2 and it's recommended practice still.

alvations replied:
fixed.

stevenbird commented on nltk/translate/stack_decoder.py:
```diff
@@ -0,0 +1,143 @@
+    
+    :type stack: list
+    :para stack: a stack is a list of `Hypotheses`.
+    
```
@alvations, please remove the blank lines between the parameter declarations, for consistency with the rest of the toolkit.

alvations replied:
fixed.

stevenbird commented on nltk/translate/stack_decoder.py:
```diff
@@ -0,0 +1,143 @@
+
+def hypothesis_to_translation(hypothesis):
+    """
+    This modules recursively iterates through the nodes in a `Hypothesis` 
```
This sentence is ungrammatical.

alvations replied:
Fixed.

stevenbird commented:
Thanks @alvations – great work! I've asked for some minor clean-ups, and then I think it's ready to merge.

alvations commented:
@stevenbird , I've fixed the clean-ups requested. Hopefully I didn't miss anything.

alvations commented:
Would it be okay if I present the current state of the translate and align module to the public in PyCon? We have a poster accepted in PyCon, http://goo.gl/kG8kOl

stevenbird commented:
Sure, that would be great!

Sorry for the delay. I just pulled this into my own source tree to test and see that it doesn't work with Python 3. Please add the `from __future__ import print_function`, and change `print` to be a function not a statement, and test with Python 3.

alvations commented:
Sorry for the slip on the tests. The code should work with python 3 now.

stevenbird commented:
Thanks @alvations. We're getting there, but I want to get the structure right. We have MT-related content in the `align` and `translate` packages, and the `translate` package contains material that belongs in the `model` package that I want to reinstate soon (cf #800). So please bear with me.

alvations commented:
@stevenbird , no problem. I would really want the `model` package to work too, then ngram extractions and estimations is possible for MT.

stevenbird commented:
Thanks @alvations. Once we have #800 properly merged, I'd be very grateful if you would work out how to use (or extend) the `model` package.

stevenbird commented:
Sorry for the delay here, @alvations. I'm hoping for progress re #800.

alvations commented:
@stevenbird , no problem about the delay.

stevenbird commented:
Things are moving forward with the language modelling package now (please see the model branch). You might like to try it out?

alvations commented:
**Questions/Suggestions about the model package**
- Is that in the [model](https://github.com/nltk/nltk/blob/model/nltk/model/ngram.py) branch? 
- What is `teardown_module()` doing? Is it used to dump the corpus in buffer after training the language model? Wouldn't that do so globally?
- Also a NgramModel dump to `arpa` textfile format would be helpful. Although I would prefer to save it in pickle binaries. Or maybe allow both textfile and pickle dump. 
- Could the padding be set to default `<s>` and `<\s>` like most language model instead of `''`? It's because `if "":` is False and it's hard to do checks when other packages uses the `NgramModel._lpad`. But the risk is that `<s>` might be retokenized to `['<', 's', '>']` if users/devs dumps `NgramModel.train` to strings and then re-read and retokenize it. But the chances of people retokenizing `NgramModel.train` should be quite rare.
- A dictionary of number of unique ngrams of different orders should be kept so that users can access it to perform renormalization or other functions that needs `len(no. of unigrams)`,  `len(no. of bigrams)`, etc. 
- Unbounded vs bounded method for entropy and perplexity, see below.
- Should the `nltk.translate.util.read_translation_model()` go to `nltk.model`? Should I create a different model abstract class for a phrase translation model? How about tree-based translation model, different or same abstract class?
- Can I contribute to add the dump to arpa file function in `model`? If so, what should we name it? Is there a convention that can do something like `nltk.corpus.dump_to_text`?
- Last important question, pardon my "noobness" in git development, should I fork a different branch from `model` and then work on the `translate` package and dump my current branch. Or should I continue on the `develop` branch?

Perplexity and entropy could be an unbound method where the user can do:

```
x = NgramModel(xtext)
y = NgramModel(ytext)
model.perplexity(x, y)
```

currently, i think one has to do:

```
x = NgramModel(xtext)
y = NgramModel(xtext)
x.perplexity(y.train)
```

Maybe we should allow both. It's sort of like the `wn.path_similarity(x,y)` vs `x.path_similarity(y)`

---

`nltk.translate` TODOs
- port `nltk.translate.util.read_lang_model()` to `nltk.model.util.read_arpa()`
- change :return: of `read_lang_model()` to `nltk.model.NgramModel`

---

Sorry for the multiple questions.

alvations commented:
About the git question, i've created a `translate` branch in my fork and i'll merge back with my `develop` branch once the `model` merges with `develop`.

alvations commented:
@stevenbird , @Copper-Head, the model NgramModel is breaking due to the smoothing:

[in]:

```
from nltk.model import NgramModel
from nltk.corpus import brown
from nltk.probability import LidstoneProbDist
est = lambda fdist, bins: LidstoneProbDist(fdist, 0.2)
lm = NgramModel(3, brown.words(categories='news'), estimator=est)
```

[out]:

```
Traceback (most recent call last):
  File "/home/alvas/git/nltk/test_model.py", line 15, in <module>
    lm = NgramModel(3, brown.words(categories='news'), estimator=est)
  File "/home/alvas/git/nltk/nltk/model/ngram.py", line 121, in __init__
    self._probdist = estimator(cfd, *estimator_args, **estimator_kwargs)
TypeError: <lambda>() takes exactly 2 arguments (1 given)
```

Even without estimator the default estimator is not working. Also the docstring in `_estimator` writes GoodTuring but it's calling lidstone =(

stevenbird commented:
@Copper-Head are you able to investigate this problem please?

stevenbird commented:
@alvations can you see what happens if you set `_estimator` to GoodTuring?

alvations commented:
@stevenbird, @Copper-Head, the `_estimator` runs with GoodTuring and Lidstone. The docstring in the `NgramModel` should have been:

```
>>> from nltk.corpus import brown
>>> from nltk.probability import LidstoneProbDist
>>> est = lambda fdist : LidstoneProbDist(fdist, 0.2)
>>> lm = NgramModel(3, brown.words(categories='news'), estimator=est)
>>> lm
```

The `est = lambda fdist : LidstoneProbDist(fdist, 0.2)` works  with the NgramModel without the returning the bins. With the bins, it breaks.

But it runs only for py2.x but it breaks in py3.x because of the `ConditionalFreqDist.iterkeys()`. 

I think `ConditionalFreqDist` would need something like `six.iterkeys()` to override the `dictionary.iterkeys()` so that it's supported both in py2.x and py3.x.

```
from nltk.model import NgramModel
from nltk.corpus import brown
from nltk.probability import LidstoneProbDist, SimpleGoodTuringProbDist


#est = lambda fdist, bins: LidstoneProbDist(fdist, 0.2)
est2 = lambda fdist: SimpleGoodTuringProbDist(fdist)

corpus = brown.words(categories='news')[:100]
lm = NgramModel(3, corpus, estimator=est2)

# Testing prob
print (corpus[8], corpus[9], corpus[12] )
print (lm.prob(corpus[12], [corpus[8], corpus[9]]) )
print


print (corpus[8], corpus[9], corpus[22] )
print (lm.prob(corpus[22], [corpus[8], corpus[9]]) )
print

print (corpus[8], corpus[9], corpus[10] )
print (lm.prob(corpus[10], [corpus[8], corpus[9]]) )
print

# Testing logprob
print (corpus[8], corpus[9], corpus[12] )
print (lm.logprob(corpus[12], [corpus[8], corpus[9]]) )
print


print (corpus[8], corpus[9], corpus[22] )
print (lm.logprob(corpus[22], [corpus[8], corpus[9]]) )
print

print (corpus[8], corpus[9], corpus[10] )
print (lm.logprob(corpus[10], [corpus[8], corpus[9]]) )
print
```

[out]:

```
alvas@ubi:~/git/nltk$ python test_model.py
(u'investigation', u'of', u'primary')
0.0124909679922

(u'investigation', u'of', u'took')
0.00400419641542

(u'investigation', u'of', u"Atlanta's")
0.0

alvas@ubi:~/git/nltk$ python3 test_model.py
Traceback (most recent call last):
  File "test_model.py", line 20, in <module>
    lm = NgramModel(3, corpus, estimator=est2)
  File "/home/alvas/git/nltk/nltk/model/ngram.py", line 129, in __init__
    **estimator_kwargs)
  File "/home/alvas/git/nltk/nltk/model/ngram.py", line 141, in __init__
    for word in self._words_following(ctxt, cfd):
  File "/home/alvas/git/nltk/nltk/model/ngram.py", line 162, in _words_following
    for ctxt, word in cond_freq_dist.iterkeys():
AttributeError: 'ConditionalFreqDist' object has no attribute 'iterkeys'
```

For GoodTuring, it seems like three's a bug **the correct ngram is getting probability of 0.0 instead of 1.0 or ~0.999.** when retrieving `LanguageModel.prob(word, context)`. When getting logprob it goes into math domain error because of Log 0. 

But for Lidstone, the LanguageModel was returning the right results:

```
from nltk.model import NgramModel
from nltk.corpus import brown
from nltk.probability import LidstoneProbDist, SimpleGoodTuringProbDist


est = lambda fdist: LidstoneProbDist(fdist, 0.2)

corpus = brown.words(categories='news')[:100]
lm = NgramModel(3, corpus, estimator=est)

# Testing prob
print (corpus[8], corpus[9], corpus[12] )
print (lm.prob(corpus[12], [corpus[8], corpus[9]]) )
print


print (corpus[8], corpus[9], corpus[22] )
print (lm.prob(corpus[22], [corpus[8], corpus[9]]) )
print

print (corpus[8], corpus[9], corpus[10] )
print (lm.prob(corpus[10], [corpus[8], corpus[9]]) )
print


# Testing logprob
print (corpus[8], corpus[9], corpus[12] )
print (lm.logprob(corpus[12], [corpus[8], corpus[9]]) )
print


print (corpus[8], corpus[9], corpus[22] )
print (lm.logprob(corpus[22], [corpus[8], corpus[9]]) )
print

print (corpus[8], corpus[9], corpus[10] )
print (lm.logprob(corpus[10], [corpus[8], corpus[9]]) )
print
```

[out]:

```
alvas@ubi:~/git/nltk$ python test_model.py
(u'investigation', u'of', u'primary')
0.0203835854682

(u'investigation', u'of', u'took')
0.0111183193463

(u'investigation', u'of', u"Atlanta's")
0.0106007067138

(u'investigation', u'of', u'primary')
5.61644834621

(u'investigation', u'of', u'took')
6.49091746413

(u'investigation', u'of', u"Atlanta's")
6.55969574211
```

Although with any smoothing there should never be a log 0, but i think `LanguageModel.logprob` needs to handle log 0 .

trivedigaurav commented:
I was able to reproduce @alvations's observations. Maybe it is a good idea to track that issue separately.

iliakur commented:
Hi everyone!
Sorry I've been out of touch. The downside of doing this in one's free time...
I've gone over my _fix_ for the original bug in the NgramModel code with some friends and we've come to the conclusion that my original solution was incorrect. We know what went wrong and how to rectify it, so I'll be pushing those changes soon.
That being said, I'll look over this thread as well and let you know if I think the problem is related to the stuff I've been doing or if it's a separate issue.

Thanks,
Ilia.

alvations commented:
Just like to check the status of the merge pull. Should this be killed and we'll try again after `nltk.model` is fixed? Or should we just give it up?

stevenbird commented:
This PR was inadvertently closed when I recreated the develop branch.

iliakur commented:
I'm still alive and will give this a shot this week.

stevenbird commented:
Thanks @Copper-Head, it would be great to resolve this.

iliakur commented:
@stevenbird I just want to make sure our logistics are clear.
I'm going to switch to my fork's model branch, revert the problematic commits, make some adjustments to docstrings etc. Then I recreate my repo's remote _model_ branch and push to it. In my PR, do use the _test_ branch as comparison, since you configured Jenkins for it?

iliakur commented:
@alvations I'm reading your list of suggestions/questions and wondering about the padding idea. Just to be clear I understand, if we adopt the `<s>`, `</s>` as default, the first trigam of a text `['He', 'opened', 'a', 'bottle', ...]` will look like `("<s>", "<s>", "He")`, correct?

alvations commented:
@Copper-Head ,  The padding adds the `<s>` and `<\s>` to the head and end of the sentence, e.g. `['He', 'opened', 'a', 'bottle', ...]` -> `['<s>', 'He', 'opened', 'a', 'bottle', ..., '<\s>']`. So the first trigram would be `('<s>', 'He', 'opened')`.

iliakur commented:
Hmm, wouldn't we have to change the implementation of [`nltk.util.ngrams`](https://github.com/nltk/nltk/blob/develop/nltk/util.py#L383) for what you suggest to work best?

iliakur commented:
@stevenbird the changes necessary for fixing the NgramModel (for real this time, I'm pretty sure) are sitting in a `model` branch on my fork. Let me know what your preferred logistics are for merging!

stevenbird commented:
Please submit a pull request to the `model` branch, thanks. It would be great to include tests that demonstrate correctness.

alvations commented:
@Copper-Head, thank you for the fixes! Yes, the Ngram object works now =)

I've tried with several ProbDists:

```
from nltk.model import NgramModel
from nltk.corpus import brown
from nltk.probability import LidstoneProbDist, SimpleGoodTuringProbDist, LaplaceProbDist, ELEProbDist

def _laplace_est(fdist, *estimator_args, **estimator_kwargs):
    """
    Default estimator function using a LidstoneProbDist.
    """
    # can't be an instance method of NgramModel as they
    # can't be pickled either.
    return LaplaceProbDist(fdist, *estimator_args, **estimator_kwargs)

def _ele_est(fdist, *estimator_args, **estimator_kwargs):
    """
    Default estimator function using a LidstoneProbDist.
    """
    # can't be an instance method of NgramModel as they
    # can't be pickled either.
    return ELEProbDist(fdist, *estimator_args, **estimator_kwargs)

def _goodturing_est(fdist, *estimator_args, **estimator_kwargs):
    """
    Default estimator function using a LidstoneProbDist.
    """
    # can't be an instance method of NgramModel as they
    # can't be pickled either.
    return SimpleGoodTuringProbDist(fdist, *estimator_args, **estimator_kwargs)


def test_ngram(est):
    est = _laplace_est

    corpus = brown.words(categories='news')[:1000]
    lm = NgramModel(5, corpus, estimator=est)

    # Good Ngram
    print (corpus[8], corpus[9], corpus[10] )
    print (lm.logprob(corpus[10], [corpus[8], corpus[9]]) )
    # Bad Ngram
    print (corpus[8], corpus[9], corpus[12] )
    print (lm.prob(corpus[12], [corpus[8], corpus[9]]) )
    # Worst ngram
    print (corpus[8], corpus[8], corpus[8] )
    print (lm.prob(corpus[8], [corpus[8], corpus[8]]) )
    print

test_ngram(_laplace_est)
test_ngram(_ele_est)
test_ngram(_goodturing_est)
```

If i'm not wrong only LidstoneProbDist and its subclasses can have a gamma value right?

@stevenbird, for administration purposes, could we create a package for `nltk.probability` instead of just a single python file. Currently, it's 2000+ lines of code with the different ProbDists.

iliakur commented:
Hey, @alvations I'm afraid I haven't merged in my changes just yet. I'm working on creating a doctest file for the ngram module today and will be pushing those changes in the evening. Then I'll open a PR and ping you guys.

stevenbird commented:
Thanks @Copper-Head. Yes @alvations you might be right, though there are five modules that are even bigger.

stevenbird commented:
@Copper-Head, @alvations – would you please update me on where things stand here?

alvations commented:
After looking at the `NgramModel` closely and talking to @Copper-Head , for now, I'm going to:
- Move `nltk.translate.api.LanguageModel` into `nltk.model` as a sibling class of `nltk.model.NgramModel`
- Move `nltk.translate.util.read_lang_model` into `nltk.model.util.read_lang_model` (also i'll improve how i'm reading the APRA file, it's not very efficient as it is now)
- Change the stack decoder code accordingly to access the `LanguageModel` like that of an `NgramModel`.

I'll be making the above changes soon once I get some free time.

---

**Should I create a `lm.py` in `nltk.model` or should the following be another class in `nltk.model.ngram`?**

```
from nltk.model.api import ModelI
class LanguageModel(ModelI):
  pass
```

---

I'll also like to suggest to add `pad_left_symbol` and `pad_right_symbol` to the [`NgramModel.__init__()`](https://github.com/alvations/nltk/blob/develop/nltk/model/ngram.py#L39) such that it's more consistent with existing Ngram models that uses `<s>` and `</s>` instead of [null string](https://github.com/alvations/nltk/blob/develop/nltk/model/ngram.py#L98).
